[GFC][Integration] BoxGeometry and render tree geometry need to be updated after layout.

sammygill opened the pull request:
#### 25b88b1c459a0d652211acc0c5fc20e89caa45a8
<pre>
[GFC][Integration] BoxGeometry and render tree geometry need to be updated after layout.
<a href="https://rdar.apple.com/161903602">rdar://161903602</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300118">https://bugs.webkit.org/show_bug.cgi?id=300118</a>

Reviewed by Alan Baradlay.

Now that we have a codepath for the grid layout algorithm, we need to add
some logic that handles the output of this process. This would entail
setting the BoxGeometry for each layout box associated with a grid item
as well as performing the integration step of updating the render tree
in the same way. This patch does so in three main steps:

1.) Return a set of GridItemRects along with the used track sizes from
GridLayout. A GridItemRect is a simple wrapper struct that holds the
border box geometry of the grid item alongside the margins for each
edge. This is done by taking the inline/block axis positions as a result
from alignment, used inline/block axis border box sizes from the final
layout step of grid items, and the margins resolved after we determined
the grid area sizes. Note that the position of these items is in the
coordinate space of its containing block, which is the grid area, and we
choose to map from this space to the coordinate space of the grid box as
part of GridFormattingContext. This is because we do not consider this
logic as strictly part of the grid layout process, so we let the
formatting context handle it. This is why the used track sizes are
needed as an output of layout.

2.) Set the box geometry for each grid item after we have returned from
GridLayout and have performed the step of mapping each location from the
grid area to the grid box. This is a mostly straightforward mapping
from the GridItemRect to box geometry with some small bit of math to to
go from the border box dimensions to the content box dimensions.

3.) Finally, in the integration code, take each grid item&apos;s BoxGeometry
and update the associated renderer with the same values. At this point
we also need to perform some final massaging to the formatting context
root in order to integrate properly. Namely, we need to make sure the
RenderGrid&apos;s internal grid is marked as not needing placement and also
has its order iterator properly set. This is so that

RenderGrid::paintChildren is happy!
Canonical link: <a href="https://commits.webkit.org/301074@main">https://commits.webkit.org/301074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d83dba4f661a63579d0dff5f0d53fb5c9c714bbe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124862 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44533 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35269 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131708 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76765 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/64422e4c-bef9-43dc-be05-c99488abe1d0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126739 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45235 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53102 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95018 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63043 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d9daa27c-fdf4-4b28-954d-f19a73e940a8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127816 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111669 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75576 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3e9127bd-13fa-410d-81dc-16c81ec75a55) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35025 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29826 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75184 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134376 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51704 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39508 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103499 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52121 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107884 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103261 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26296 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48621 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26908 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48723 "Hash d83dba4f for PR 51777 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51580 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57388 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50969 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54325 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52663 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

yulunwu2 commented on Source/WebCore/layout/formattingContexts/grid/GridItemRect.h:
```diff
@@ -0,0 +1,43 @@
+};
+
+} // namespace Layout
+} // namespacec WebCore
```
namespace

sammygill replied:
Fixed

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -72,9 +72,40 @@ static inline Layout::GridFormattingContext::GridLayoutConstraints constraintsFo
+        renderer->setMarginBefore(gridItemGeometry.marginBefore());
+        renderer->setMarginAfter(gridItemGeometry.marginAfter());
+        renderer->setMarginStart(gridItemGeometry.marginStart());
+        renderer->setMarginAfter(gridItemGeometry.marginAfter());
```
setMarginEnd() ?

sammygill replied:
Oops good catch! Fixed

yulunwu2 commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
```diff
@@ -150,10 +167,34 @@ PlacedGridItems GridFormattingContext::constructPlacedGridItems(const GridAreas&
 }
 
-const BoxGeometry GridFormattingContext::geometryForGridItem(const ElementBox& gridItem) const
+const BoxGeometry GridFormattingContext::geometryForGridItem(const ElementBox& layoutBox) const
```
do we want this to be a ref?

sammygill replied:
Yeah fixed!

yulunwu2 commented on Source/WebCore/layout/formattingContexts/grid/GridItemRect.h:
```diff
@@ -0,0 +1,43 @@
+
+LayoutRect m_borderBoxRect;
+RectEdges<LayoutUnit> m_margins;
+const GridAreaLines gridAreaLines;
```
m_gridAreaLines

sammygill replied:
Fixed

yulunwu2 reviewed (commented):
added some comments, overall LGTM

alanbaradlay commented on Source/WebCore/layout/formattingContexts/grid/GridItemRect.h:
```diff
@@ -0,0 +1,43 @@
+LayoutRect m_borderBoxRect;
+RectEdges<LayoutUnit> m_margins;
+const GridAreaLines m_gridAreaLines;
+const CheckedRef<const ElementBox> m_layoutBox;
```
missing indentation.

alanbaradlay replied:
struct members don't need `m_ ` prefix

alanbaradlay commented on Source/WebCore/layout/formattingContexts/grid/GridItemRect.h:
```diff
@@ -0,0 +1,43 @@
+
+LayoutRect m_borderBoxRect;
+RectEdges<LayoutUnit> m_margins;
+const GridAreaLines m_gridAreaLines;
```
maybe replace it with a more descriptive name?

sammygill committed:
[GFC][Integration] BoxGeometry and render tree geometry need to be updated after layout.
rdar://161903602
https://bugs.webkit.org/show_bug.cgi?id=300118

Reviewed by Alan Baradlay.

Now that we have a codepath for the grid layout algorithm, we need to add
some logic that handles the output of this process. This would entail
setting the BoxGeometry for each layout box associated with a grid item
as well as performing the integration step of updating the render tree
in the same way. This patch does so in three main steps:

1.) Return a set of GridItemRects along with the used track sizes from
GridLayout. A GridItemRect is a simple wrapper struct that holds the
border box geometry of the grid item alongside the margins for each
edge. This is done by taking the inline/block axis positions as a result
from alignment, used inline/block axis border box sizes from the final
layout step of grid items, and the margins resolved after we determined
the grid area sizes. Note that the position of these items is in the
coordinate space of its containing block, which is the grid area, and we
choose to map from this space to the coordinate space of the grid box as
part of GridFormattingContext. This is because we do not consider this
logic as strictly part of the grid layout process, so we let the
formatting context handle it. This is why the used track sizes are
needed as an output of layout.

2.) Set the box geometry for each grid item after we have returned from
GridLayout and have performed the step of mapping each location from the
grid area to the grid box. This is a mostly straightforward mapping
from the GridItemRect to box geometry with some small bit of math to to
go from the border box dimensions to the content box dimensions.

3.) Finally, in the integration code, take each grid item's BoxGeometry
and update the associated renderer with the same values. At this point
we also need to perform some final massaging to the formatting context
root in order to integrate properly. Namely, we need to make sure the
RenderGrid's internal grid is marked as not needing placement and also
has its order iterator properly set. This is so that

RenderGrid::paintChildren is happy!
Canonical link: https://commits.webkit.org/301074@main
